cmd/cstor-volume-mgmt/controller/volume-controller/new_volume_controller.go

```diff
@@ -0,0 +1,143 @@
+
+	apis "github.com/openebs/maya/pkg/apis/openebs.io/v1alpha1"
+	clientset "github.com/openebs/maya/pkg/client/clientset/versioned"
+	openebsScheme "github.com/openebs/maya/pkg/client/clientset/versioned/scheme"
+	informers "github.com/openebs/maya/pkg/client/informers/externalversions"
+)
+
+const volumeControllerName = "CStorVolume"
+
+// CStorVolumeController is the controller implementation for CStorVolume resources.
+type CStorVolumeController struct {
+	// kubeclientset is a standard kubernetes clientset
+	kubeclientset kubernetes.Interface
+
+	// clientset is a openebs custom resource package generated for custom API group.
+	clientset clientset.Interface
+
+	// cStorVolumeSynced is used for caches sync to get populated
+	cStorVolumeSynced cache.InformerSynced
+
+	// deletedIndexer holds deleted resource to be retrived after workqueue
+	deletedIndexer cache.Indexer
+
+	// workqueue is a rate limited work queue. This is used to queue work to be
+	// processed instead of performing it as soon as a change happens. This
+	// means we can ensure we only process a fixed amount of resources at a
+	// time, and makes it easy to ensure we are never processing the same item
+	// simultaneously in two different workers.
+	workqueue workqueue.RateLimitingInterface
+	// recorder is an event recorder for recording Event resources to the
+	// Kubernetes API.
+	recorder record.EventRecorder
+}
+
+// NewCStorVolumeController returns a new instance of CStorVolume controller
+func NewCStorVolumeController(
+	kubeclientset kubernetes.Interface,
+	clientset clientset.Interface,
+	kubeInformerFactory kubeinformers.SharedInformerFactory,
+	cStorInformerFactory informers.SharedInformerFactory) *CStorVolumeController {
+
+	// obtain references to shared index informers for the cStorVolume resources
+	cStorVolumeInformer := cStorInformerFactory.Openebs().V1alpha1().CStorVolumes()
+
+	openebsScheme.AddToScheme(scheme.Scheme)
+
+	// Create event broadcaster to receive events and send them to any EventSink, watcher, or log.
+	// Add NewCstorVolumeController types to the default Kubernetes Scheme so Events can be
+	// logged for CstorVolume Controller types.
+	glog.V(4).Info("Creating event broadcaster")
+	eventBroadcaster := record.NewBroadcaster()
+	eventBroadcaster.StartLogging(glog.Infof)
+
+	// StartEventWatcher starts sending events received from this EventBroadcaster to the given
+	// event handler function. The return value can be ignored or used to stop recording, if
+	// desired. Events("") denotes empty namespace
+	eventBroadcaster.StartRecordingToSink(&typedcorev1.EventSinkImpl{Interface: kubeclientset.CoreV1().Events("")})
+	recorder := eventBroadcaster.NewRecorder(scheme.Scheme, corev1.EventSource{Component: volumeControllerName})
+
+	controller := &CStorVolumeController{
+		kubeclientset: kubeclientset,
+		clientset:     clientset,
+		deletedIndexer: cache.NewIndexer(cache.DeletionHandlingMetaNamespaceKeyFunc,
```

Comment:
same comment, deletedIndexer can be overlooked if it is being used with latest finalizer approach